api/admin/controller.py

```diff
@@ -388,6 +419,68 @@ def resolve_complaints(self, data_source, identifier):
             return COMPLAINT_ALREADY_RESOLVED
         return Response("", 200)
 
+    def update_genres(self, data_source, identifier):
+        pool = self.load_licensepool(data_source, identifier)
+        if isinstance(pool, ProblemDetail):
+            return pool
+        work = pool.work
+        staff_data_source = DataSource.lookup(self._db, DataSource.LIBRARY_STAFF)
+        new_genres = flask.request.form.getlist("genres")
+
+        current_staff_classifications = work.classifications_with_genre() \
+            .filter(Classification.data_source_id == staff_data_source.id) \
+            .all()
+        current_staff_genres = [
+            c.subject.genre.name 
+            for c in current_staff_classifications 
+            if c.subject.genre
+        ]
+
+        # make sure all new genres are legit
+        for name in new_genres:
+            genre, is_new = Genre.lookup(self._db, name)
+            if not isinstance(genre, Genre):
+                return GENRE_NOT_FOUND
+            if work.fiction != genres[name].is_fiction:
+                return INCOMPATIBLE_GENRE
+
+        # delete existing staff classifications for genres that aren't being kept
+        for c in current_staff_classifications:
+            if c.subject.genre.name not in new_genres:
+                self._db.delete(c)
+
+        # add new staff classifications for new genres
+        for genre in new_genres:
+            if genre not in current_staff_genres:
+                classification = work.primary_edition.primary_identifier.classify(
+                    data_source=staff_data_source,
+                    subject_type=Subject.SIMPLIFIED_GENRE,
+                    subject_identifier=genre,
+                    weight=WorkController.STAFF_WEIGHT
+                )
+
+        # add NONE classification if we aren't keeping any genres
+        staff_classifications = work.classifications_with_genre() \
+            .filter(Classification.data_source_id == staff_data_source.id) \
+            .all()
```

Comment:
This should be the same as checking in len(new_genres) is 0, which would save a database query. Maybe it's safer to do this to make sure?